src/main/java/tc/oc/pgm/freeze/FreezeManager.java

```diff
@@ -0,0 +1,124 @@
+package tc.oc.pgm.freeze;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.bukkit.ChatColor;
+import org.bukkit.Material;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemFlag;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import tc.oc.pgm.AllTranslations;
+import tc.oc.pgm.Config;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.Permissions;
+import tc.oc.pgm.api.chat.Sound;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.match.ObservingParty;
+
+/** Controls player freezing and those in freeze mode */
+public class FreezeManager {
```

Comment:
No need for a `FreezeManager`. I would rather implement a `MatchPlayer#setFrozen(boolean)` method.

Reply:
setFrozen already exists, but it's for player movement prevention (like when you die), not for punishment-wise freezing

Reply:
Wait, then why can't we use that code?

Reply:
@Electroid well, they should be using it. That is the "low-level" way to prevent the player from moving, then everyting else that goes ontop knowing who's frozen, being able to unfreeze, that stuff, needs to be handled outisde. Just because someone died doesn't mean you can run /unfreeze on that player

Reply:
What I'm trying to understand is why is `FreezeListener` required if we already have a freezing mechanism with `isFrozen`? (Aside from the title and UI stuff that says you're frozen).

Reply:
`isFrozen` is just the player state of not being able to move, that doesn't know why that player was put on frozen mode.

If you would use that as the punishment flag, you could run /unfreeze on dead people and pgm would be like "Unfrozen player", when in reality, that player wasn't frozen (the punishment-freeze).

As to why the listener exists, is because it still has to prevent you from throwing items, or placing blocks, that `isFrozen` doesn't handle.
 When you're dead what handles no-block-placing is the fact that you are on the `Spawning` state, not the fact that you're frozen. When you're frozen in-game, it has to prevent block placing even if your player state is `Alive`.